### ISSUE TITLE
File upload over https fails signature verification

### ISSUE THREAD

**Issue by danilogiacomi** [livewire/livewire#1216]
### Set up livewire file upload, then use https (e.g. with ngrok) the upload fails with the request to

Request URL: https://XYZ.ngrok.io/livewire/upload-file?expires=1593877850&signature=0c0b334969966e642854a730fdb78accd3c7332648291fc2449530f15cea7035
Request Method: POST
Status Code: 401 Unauthorized

Following the flow with the debugger the problem arises, in the vendor/livewire/livewire/src/Controllers/FileUploadHandler.php, in the handle method.

In particular in the 

abort_unless(request()->hasValidSignature(), 401); line.

Commenting it out works around the problem (the uploads works) but of course is not optimal,

**Context**
- Livewire version: 1.3.0
- Laravel version: 7.10.0
- Browser: [Firefox, Vivaldi]

**Comment by selfeky** [comment]
I have the same problem!
@danilogiacomi did you find a solution?

**Comment by danilogiacomi** [comment]
> I have the same problem!
> @danilogiacomi did you find a solution?

No, at the moment I've commented that line out, but that's not a solution at all...

**Comment by dennisvandalen** [comment]
@danilogiacomi @selfeky 

This is because the traffic is proxied. You can set the trusted proxy in the `TrustProxies` middleware as per the Laravel docs.

The simplest way is to trust all proxies if you don't know them:
```php
protected $proxies = '*';
```

If you're behind Cloudflare or some kind of load balancer this also applies.

**Comment by selfeky** [comment]
@dennisvandalen Thanks! It works!

**Comment by alexpgates** [comment]
In my case, the upload _was_ working, but the `temporaryUrl() ` method wasn't working. I'm on a Forge-provisioned Ubuntu 20.04 server. 

Adjusting the trusted proxies didn't work, but commenting out:

`abort_unless(request()->hasValidSignature(), 401);`

in `/vendor/livewire/livewire/src/Controllers/FilePreviewHandler.php` worked.

Not ideal, but wanted to add this in case anyone else is having issues with the `temporaryUrl()` method.

**Comment by bagwaa** [comment]
@alexpgates thanks for this, this whole issue cropped up for us when we were behind a load balancer and forcing https, just can't seem to get a valid signature under those circumstances.

Ended up commenting out the above code in the vendor directory as @alexpgates mentioned, but also on FileUploadHandler which has a similar check.

**Comment by gillidandaweb** [comment]
It also fails with the 401 if the upload takes more than 5 minutes.  Maybe this could be configurable?  Currently it's hard coded in `GenerateSignedUploadUrl`:

forLocal:

```
        return URL::temporarySignedRoute(
            'livewire.upload-file', now()->addMinutes(5)
        );
```

forS3:

```
        $signedRequest = $adapter->getClient()->createPresignedRequest(
            $command,
            '+5 minutes'
        );
```

**Comment by mokhosh** [comment]
Same problem using cloudflare for ssl

**Comment by mvnobrega** [comment]
Hi,

In my case the same route but the issues is about request the url in http over https. I receve the message:

This request has been blocked; the content must be served over HTTPS.

**Comment by coolsam726** [comment]
> @danilogiacomi @selfeky
> 
> This is because the traffic is proxied. You can set the trusted proxy in the `TrustProxies` middleware as per the Laravel docs.
> 
> The simplest way is to trust all proxies if you don't know them:
> 
> ```
> protected $proxies = '*';
> ```
> 
> If you're behind Cloudflare or some kind of load balancer this also applies.

I am running my app in Docker, behind an apache proxy. I have tried this but it doesn't work for me. @calebporzio please shed some light on this issue. Thanks.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by ijpatricio** [comment]
REOPEN

**Comment by ijpatricio** [comment]
@coolsam726 I'm having the exact same scenario: docker, behind reverse proxy, and using a subdomain

any chance that you solved it further, and still didn't post more updates here??

**Comment by ijpatricio** [comment]
so i can confirm, "solution" here https://forum.laravel-livewire.com/t/file-upload-over-https-fails-signature-verification/1242
makes it work!

TLDR = comment out `abort_unless(request()->hasValidSignature(), 401)`
in `vendor/livewire/livewire/src/Controllers/FileUploadHandler.php` @ `handle` method.

**Comment by ijpatricio** [comment]
Thank you @danilogiacomi ! Sorry that just now realised you must be Dan from the livewire forum, so it was you opening the issue originally!

**Comment by ijpatricio** [comment]
So, I tried to change the behaviour as a quick dirty fix

![image](https://user-images.githubusercontent.com/26031459/99068301-2aed6f80-25a4-11eb-9284-0c5ea13c7da7.png)

but the `Livewire\Controllers\FileUploadHandler@handle` is using request from the helper `request()`

if it would be injeting on the constructor, we could leverage contextual binding as above...

![image](https://user-images.githubusercontent.com/26031459/99068426-68ea9380-25a4-11eb-84d9-9b7a50037e81.png)

@calebporzio i would all in to try and solve things the proper way, but i have no idea where to start on the signature thing. any pointers?

**Comment by coolsam726** [comment]
> @coolsam726 I'm having the exact same scenario: docker, behind reverse proxy, and using a subdomain
> 
> any chance that you solved it further, and still didn't post more updates here??

I did not manage to solve it. The issue is actually deeply rooted in Laravel's UrlGenerator especially when you are accessing the url through the request object.

If you say you are running your project in a __subdomain__ then in your case a solution might be possible.

When you are running behind a load-balancer, the most common problems are resolving a sub-directory url e.g mydomain.com/myapp/ AND applying https to the source data (Using `X-Forwarded-Proto` Header).

If you are using apache for proxying, enable the headers module (`sudo a2enmod headers`) on your host and add the following directive in your apache config file:
```
RequestHeader set X-Forwarded-Proto https
```
Also ensure you have APP_URL set correctly. I hope this will do some good.

**Comment by angelofan** [comment]
I tried all the solutions, but it still didn't work.
I use https and use with CDN

**Comment by johannesobier** [comment]
I've this problem, solvable by uncommenting the mentioned line, and I'm not using a proxy. But a nginx-self-signed certificate on my development VM with nginx. On the other hand this problem doesn't occur with another project with IIS (and an IIS-self-signed certificate) on a Windows Server 2012 R2 VM. Strange...

**Comment by breadthe** [comment]
I ran into the same issue when I attempted to upload files > 64MB, but managed to solve it a different way, after trying some of the suggestions here.

It should have been obvious, except that I tested in Chrome which showed `Failed to load resource: net::ERR_CONNECTION_CLOSED - HTTP Error 401 Unauthorized`, and not the actual error returned by my server.

Switching to Firefox, I saw an nginx `413 Request Entity Too Large` error, which told me right away that I had configured file uploaded size too low in nginx.

I changed:
- `client_max_body_size 200M;` in `/etc/nginx/sites-enabled/sub.domain.tld`
- `post_max_size = 200M` and `upload_max_filesize = 200M` in `/etc/php/7.4/fpm/php.ini`

Now am I able to upload files up to 200M (duh!).

So if you can upload files only up to a certain size before getting the `HTTP Error 401 Unauthorized`, check your nginx + PHP configuration before trying other measures.

**Comment by coolsam726** [comment]
> It looks like this thread has multiple different problem domains contained within it.
> 
> Let's keep the discussion to uploads not working with https behind proxies/sub-directories.
> 
> The fix suggested earlier where we just comment out that "abort_unless" line doesn't seem safe to me. I don't have my head around the source at the moment, but I would love it if someone who understands the problem proposes a solution or list of potential solutions to clarify the issue. Thanks!


@calebporzio I couldn't agree more. Commenting out "abort_unless" is simply saying goodbye to Laravel's CSRF Protection for your project and is not wise at all.

As I had [mentioned here earlier](https://github.com/livewire/livewire/issues/1216#issuecomment-726770588), the issue at hand is actually NOT a livewire issue, but is deeply rooted in Laravel, and the fact that Laravel DOES NOT officially support running projects in sub-directories. In fact any issue opened related to sub-directory issues is not given attention by the Laravel maintainers.

**The issue:**
* When running behind a load-balancer and proxying through a subpath, Laravel still resolves routes using the root domain. e.g under the hood `/my-subpath/users` would end up being resolved as `/users` by the UrlGenerator. As a result, the CSRF verifier tries to compare the two urls and it encounters a conflict, hence returns **Invalid Signature**. 
* The other scenario is where you are running behind a load-balancer without ssl (e.g `http:127.0.0.1:8002` proxied as `https://mydomain.com`. In this case the same verifier still returns invalid signature, unless a header is passed to the proxy to show that this app should run via https.
* The final scenario is what is addressed by @dennisvandalen [In this comment](https://github.com/livewire/livewire/issues/1216#issuecomment-667944026).

**The solutions:**
1. DO NOT run your project under a subpath of a domain. Always try to use a sub-domain instead of a subpath
2. Configure TrustedProxies according to @dennisvandalen's answer.
3. If you are using Apache for proxying, enable headers in the config and add the following directive:
```
RequestHeader set X-Forwarded-Proto https
```
Someone could help with the Nginx equivalent of that.

**NB:** I have suffered this problem for a while but finally those three resolves helped me solve it. See if they could help you if you have a similar issue.

**Comment by calebporzio** [comment]
@coolsam726, that explanation is so perfect. Thanks so much for putting that tother.

For others viewing this issue, I'm closing this due to the reasons mentioned about.

Thanks again!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by eduardichim** [comment]
REOPEN

**Comment by eduardichim** [comment]
I think this is a Livewire issue, I don't know how to fix it but I can tell you what I've found

On any route that I've defined doing `request()->query()` will return the query parameters on url for some strange reason on the `livewire.preview-file` route this doesn't happen.

So I'm going to explain as best as I can the process and how to replicate.

Add the following code to one of your controller:

`
URL::temporarySignedRoute(
            'livewire.preview-file',
            now()->addMinutes(30),
            [
                'filename' => 'somefile.png'
            ]
        )
`
This code is taken from `TemporaryUploadedFile@temporaryUrl`


Now if you go to `Illuminate/Routing/UrlGenerator@signedRoute` and on line 336 (above the return) hit it with a `dd($this->route($name, $parameters, $absolute))` you will see that the url contains the expires parameter (timestamp generated by `TemporaryUploadedFile@temporaryUrl`). 

Going to the url generated `/livewire/preview-file/somefile.png?expires={timestamp}&signature={signature}` will access `Livewire\Controllers\FilePreviewHandler@handle` (where people said to remove the `abort_unless(...)` etc. - I think this is wrong) we can see it to use the `request()->hasValidSignature()` method which calls `$this->hasCorrectSignature(...)` method.

Going to `Illuminate/Routing/UrlGenerator@hasCorrectSignature` and on line 394 (above the signature being created) hitting it with a `dd($original)` you will see that the expires is **NO** where to be found... Futher investigation shows that if on line 390 (above the definition of `$original`) doing a `dd($request->query())` this will return empty... going on `Livewire\Controllers\FilePreviewHandler@handle` and in handle above the `abort_unless(...)` doing a `dd(request()->query())` is empty WTH!!

In conclusion it will always be 401 because the URL being signed are never the same.
This might not be about proxy, or http headers is just the `request()->query()` empty...

Going to `LivewireServiceProvider` and doing `dd(request()->query())` as the first thing on `boot` or `register`is still empty, I've never seen this happening on any of the other packages

**Comment by eduardichim** [comment]
If anyone is asking I'm using https://laravel.com/docs/8.x/deployment#nginx config on nginx...

I just found out this is for files only `.png?param=value`...

**Comment by johannesobier** [comment]
I don't know wether the nginx variable $query_string really changed to $args. I made a nginx setup with a tutorial that also uses $query string but found out that on propably only newer nginx versions you have to use $args. Maybe the reason behind this is something different, I'm not aware of. 

So instead of
try_files $uri $uri/ /index.php?$query_string
try this:
try_files $uri $uri/ /index.php?$args

In my case this solved the problem. The reason why the signature verification failed was simply that with $query_string in the site's config, the query string simply was lost, so that it seemed (!) to be empty. And therefore it resulted in the signature verification failure.

**Comment by johannesobier** [comment]
After a little more research it seems, that when using $query_string in conjunction with auth, the ? gets urlencoded. So before any processing the main url is tampered and the query string itself is lost:
https://trac.nginx.org/nginx/ticket/761
I don't know why this doesn't happen with $args. I only know when using $args it works even in conjunction with auth.

**Comment by eduardichim** [comment]
@johannesobier 

It seems is working with both $args and $query_string, I'm running nginx 1.10.3

Not my problem is with this...

`
    location ~* \.(?:jpg|jpeg|png|webp|gif|ico|css|js|woff2|svg)$ {
        expires 30d;
        access_log off;
        add_header Pragma public;
        add_header Cache-Control "public";
    }
`

For some reason if I comment this out it will work lovely...

**Comment by eduardichim** [comment]
Hi everyone,

It seems that in order to work the cache for static files while also keeping your cache settings do not use a location but instead do something like this:

```


    location / {

        if ($request_uri ~* ".(jpg|jpeg|png|webp|gif|ico|css|js|woff2|svg)$") {
            expires 30d;
            access_log off;
            add_header Pragma public;
            add_header Cache-Control "public";
            break;
        }

        try_files $uri $uri/ /index.php?$query_string;
    }
```

**Comment by nielskramerr** [comment]
Hi all,

Was having the same issue in combination with NGINX. We found that adding another piece to your NGINX config did the trick, for some reason the livewire request didn't get into the default, for us.
```
location / {
}
```
But adding the config below added the signature query string to the response and resulted in a working temporaryUrl
```  
location ^~ /livewire {
    try_files $uri $uri/ /index.php?$query_string;
}
```

**Comment by austenc** [comment]
Sounds like you folks have found a solution! Please reopen if that is not the case. Thanks!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by mvnobrega** [comment]
REOPEN

**Comment by mvnobrega** [comment]
I try the nginx config but now I have one message about POST is not supported for this route.

**Comment by mvnobrega** [comment]
Work for me include the location on nginx:

location ^~ /livewire {
    try_files $uri $uri/ /index.php?$query_string;
}

Force HTTPS in laravel include this line in AppServiceProvider on boot method:

        if (App::environment('production', 'development'))
        { 
            URL::forceScheme('https');
        }

Include this line on TrustProxies:

       protected $proxies = '*';

And modify php.ini for remove tmpfile() of disable_functions.

**Comment by klrkdekira** [comment]
Just to ping somehow setting TrustProxies with the following doesn't work for me, it resulted with route generated with a `:80` suffix. 

    protected $proxies =  '*';

Or I might be missing something here.

**Comment by up2top** [comment]
I faced the file upload signature verification problem when uploading user avatars via Livewire component in Laravel. The preview images of freshly uploaded files were not displayed by the `$file->temporaryUrl()` function.

Setting `TrustProxies` middleware to trust all the proxies didn't solve the problem.

Earlier I added the following lines to the Nginx site configuration of the Forge service to improve browser caching of static assets:

```
    location ~*  \.(jpg|jpeg|png|gif|ico|css|js|pdf)$ {
        expires 7d;
    }
```
	
Removing all image extensions from this setting fixed the problem. So, the final working setting is the following:

```
    location ~*  \.(ico|css|js|pdf)$ {
        expires 7d;
    }
```
	
Now preview images of uploaded files are displayed correctly. As for caching of images, I moved all of them to the separate CDN server.

**Comment by corbinjurgens** [comment]
I may have spotted the cause for my situation.

EVEN if you use 
```php
\URL::forceScheme( request()->header('X-Forwarded-Proto', 'https') );
```
or similar in your AppServiceProvers boot, 

the requsests url will still technically be http, by that I mean calling request()->url(), which is how hasValidSignature is verified.

However I haven't find a quick fix to also force requests to be the correct schema

**Comment by lnfel** [comment]
> @danilogiacomi @selfeky
> 
> This is because the traffic is proxied. You can set the trusted proxy in the `TrustProxies` middleware as per the Laravel docs.
> 
> The simplest way is to trust all proxies if you don't know them:
> 
> ```
> protected $proxies = '*';
> ```
> 
> If you're behind Cloudflare or some kind of load balancer this also applies.

I am using cloudflare and this one just worked, it made sense to me the first time I read Dennis' explanation.

Regarding caching, I prefer to avoid doing it on server level (Apache on my case). Either do it using redis or the default laravel file cache. Google's page insights would still yell at me by doing this, since they kinda expect the server level cache policy. So I add this simple snippet to my .htaccess file:

```
# Send Requests To Front Controller...
Rewrite.... ommitted lines for brevity
. 
.
.
<FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js|webp|png|jpg|ico)$">
    Header set Access-Control-Allow-Origin "*"
    Header set Cache-Control "max-age=31536000, public"
</FilesMatch>
```
By setting cache control to 1 year, page insights test result would pass.
There might be better way to this but it is much more easier for me to handle caching on app level.

**Comment by lnfel** [comment]
Apparently the trusted proxies answer did not work.
What worked for me is adding Content-Security-Policy upgrade-insecure-requests, simply add it so the head element below title tag:
```
<meta http-equiv="Content-Security-Policy" content="upgrade-insecure-requests">
```
This will upgrade any request that is sent over http to https instead, note by default most browsers to this automatically if the page being viewed is served over https but sometimes it would overlook other requests over http and simple do nothing, just like livewire upload. In this case we tell the browser to upgrade-insecure-requests all request instead.

sources: 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/upgrade-insecure-requests
https://web.dev/fixing-mixed-content/
